File: backend/app/db/models/playbook.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class PlaybookPhase(Base):
    __tablename__ = "playbook_phases"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    phase_number: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str] = mapped_column(String, nullable=False)
    day_start: Mapped[int] = mapped_column(Integer, nullable=False)
    day_end: Mapped[int] = mapped_column(Integer, nullable=False)
    outcome_target: Mapped[str] = mapped_column(Text, nullable=False)

    weeks: Mapped[list["PlaybookWeek"]] = relationship(
        "PlaybookWeek", back_populates="phase", cascade="all, delete-orphan", order_by="PlaybookWeek.week_number"
    )


class PlaybookWeek(Base):
    __tablename__ = "playbook_weeks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    phase_id: Mapped[int] = mapped_column(Integer, ForeignKey("playbook_phases.id"), nullable=False)
    week_number: Mapped[int] = mapped_column(Integer, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    day_start: Mapped[int] = mapped_column(Integer, nullable=False)
    day_end: Mapped[int] = mapped_column(Integer, nullable=False)
    outcome_target: Mapped[str] = mapped_column(Text, nullable=False, default="")
    status: Mapped[str] = mapped_column(String, nullable=False, default="pending")

    phase: Mapped["PlaybookPhase"] = relationship("PlaybookPhase", back_populates="weeks")
    tasks: Mapped[list["PlaybookTask"]] = relationship(
        "PlaybookTask", back_populates="week", cascade="all, delete-orphan", order_by="PlaybookTask.sort_order"
    )


class PlaybookTask(Base):
    __tablename__ = "playbook_tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    week_id: Mapped[int] = mapped_column(Integer, ForeignKey("playbook_weeks.id"), nullable=False)
    day_focus: Mapped[str] = mapped_column(String, nullable=False, default="")
    title: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    effort_hours: Mapped[float] = mapped_column(Float, nullable=False, default=4.0)
    priority: Mapped[str] = mapped_column(String, nullable=False, default="P1")
    status: Mapped[str] = mapped_column(String, nullable=False, default="pending")
    completion_note: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    week: Mapped["PlaybookWeek"] = relationship("PlaybookWeek", back_populates="tasks")


class PlaybookStart(Base):
    __tablename__ = "playbook_start"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    started_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=func.now())

File: frontend/src/api/playbook.ts
import client from "@/api/client";

export interface PlaybookTask {
  id: number;
  week_id: number;
  day_focus: string;
  title: string;
  description: string;
  effort_hours: number;
  priority: string;
  status: string;
  completion_note: string | null;
  completed_at: string | null;
  sort_order: number;
}

export interface PlaybookWeek {
  id: number;
  phase_id: number;
  week_number: number;
  title: string;
  day_start: number;
  day_end: number;
  outcome_target: string;
  status: string;
  tasks: PlaybookTask[];
}

export interface PlaybookPhase {
  id: number;
  phase_number: number;
  name: string;
  day_start: number;
  day_end: number;
  outcome_target: string;
  weeks: PlaybookWeek[];
}

export interface PlaybookStatus {
  day_number: number;
  current_phase: string | null;
  current_week: string | null;
  total_tasks: number;
  completed_tasks: number;
  completion_pct: number;
  started_at: string | null;
}

export interface Decision {
  question: string;
  answer: string;
}

export async function getPlaybookStatus(): Promise<PlaybookStatus> {
  const { data } = await client.get("/playbook/status");
  return data;
}

export async function getPlaybookPhases(): Promise<PlaybookPhase[]> {
  const { data } = await client.get("/playbook/phases");
  return data;
}

export async function patchPlaybookTask(
  taskId: number,
  body: { status?: string; completion_note?: string }
): Promise<PlaybookTask> {
  const { data } = await client.patch(`/playbook/tasks/${taskId}`, body);
  return data;
}

export async function createPlaybookTask(body: {
  week_id: number;
  title: string;
  description?: string;
  effort_hours?: number;
  priority?: string;
  day_focus?: string;
}): Promise<PlaybookTask> {
  const { data } = await client.post("/playbook/tasks", body);
  return data;
}

export async function getDoNotDo(): Promise<{ items: string[] }> {
  const { data } = await client.get("/playbook/do-not-do");
  return data;
}

export async function getDecisions(): Promise<{ decisions: Decision[] }> {
  const { data } = await client.get("/playbook/decisions");
  return data;
}

export async function initPlaybook(): Promise<{ seeded: boolean; phases: number }> {
  const { data } = await client.post("/playbook/init");
  return data;
}
